awx/main/migrations/0109_v370_event_bigint.py

```diff
@@ -0,0 +1,118 @@
+# Generated by Django 2.2.8 on 2020-02-21 16:31
+
+from django.db import migrations, models, connection
+
+
+def migrate_event_data(apps, schema_editor):
+    # see: https://github.com/ansible/awx/issues/6010
+    #
+    # the goal of this function is to end with event tables (e.g., main_jobevent)
+    # that have a bigint primary key (because the old usage of an integer
+    # numeric isn't enough, as its range is about 2.1B, see:
+    # https://www.postgresql.org/docs/9.1/datatype-numeric.html)
+
+    # unfortunately, we can't do this with a simple ALTER TABLE, because
+    # for tables with hundreds of millions or billions of rows, the ALTER TABLE
+    # can take *hours* on modest hardware.
+    #
+    # the approach in this migration means that post-migration, event data will
+    # *not* immediately show up, but will be repopulated over time progressively
+    # the trade-off here is not having to wait hours for the full data migration
+    # before you can start and run AWX again (including new playbook runs)
+    for tblname in (
+        'main_jobevent', 'main_inventoryupdateevent',
+        'main_projectupdateevent', 'main_adhoccommandevent',
+        'main_systemjobevent'
+    ):
+        with connection.cursor() as cursor:
+            # rename the current event table
+            cursor.execute(
+                f'ALTER TABLE {tblname} RENAME TO _old_{tblname};'
+            )
+            # create a *new* table with the same schema
+            cursor.execute(
+                f'CREATE TABLE {tblname} (LIKE _old_{tblname} INCLUDING ALL);'
+            )
+            # alter the *new* table so that the primary key is a big int
+            cursor.execute(
+                f'ALTER TABLE {tblname} ALTER COLUMN id TYPE bigint USING id::bigint;'
+            )
+
+            # recreate counter for the new table's primary key to
+            # start where the *old* table left off (we have to do this because the
+            # counter changed from an int to a bigint)
+            cursor.execute(f'DROP SEQUENCE IF EXISTS "{tblname}_id_seq" CASCADE;')
+            cursor.execute(f'CREATE SEQUENCE "{tblname}_id_seq";')
+            cursor.execute(
+                f'ALTER TABLE "{tblname}" ALTER COLUMN "id" '
+                f"SET DEFAULT nextval('{tblname}_id_seq');"
+            )
+            cursor.execute(
+                f"SELECT setval('{tblname}_id_seq', (SELECT MAX(id) FROM _old_{tblname}), true);"
+            )
+
+            # replace the BTREE index on main_jobevent.job_id with
+            # a BRIN index to drastically improve per-UJ lookup performance
+            # see: https://info.crunchydata.com/blog/postgresql-brin-indexes-big-data-performance-with-minimal-storage
+            if tblname == 'main_jobevent':
+                cursor.execute("SELECT indexname FROM pg_indexes WHERE tablename='main_jobevent' AND indexdef LIKE '%USING btree (job_id)';")
```

Comment:
@ghjm @wenottingham @jakemcdermott @matburt @chrismeyersfsu since we're effectively reinserting *all* of this data post-migration, this is a good opportunity to re-evaluate the indexes we use (as we'll be populating them from scratch on the new table we've built here).  Because this data is generally *inserted* sequentially (i.e., the `job_id` naturally tends to trend upwards in time with insertion order), I found a `BRIN` index to be *much* more efficient in the queries generated for job stdout listings.  `BRIN` indexes *are* lossy, but we don't ever *update* this data; it's set and remains constant at insertion time.  Also, it takes up **considerably** less disk space.

Here's an install w/ 100M **total** events, and one job that has 1M events attached to it.  Before:

<img width="585" alt="image" src="https://user-images.githubusercontent.com/214912/75256073-e53c0d00-57b0-11ea-806a-71601d04b390.png">

After this PR/migration are applied:

<img width="568" alt="image" src="https://user-images.githubusercontent.com/214912/75256139-fe44be00-57b0-11ea-9337-bac4fb931591.png">

Reply:
This is amazing work. Thanks for looking into this @ryanpetrello. 

>  Because this data is generally inserted sequentially (i.e., the job_id naturally tends to trend upwards in time with insertion order), I found a BRIN index to be much more efficient in the queries generated for job stdout listings. BRIN indexes are lossy, but we don't ever update this data

From what I can tell this sounds like a great match:

> The query executor is in charge of rechecking these tuples and discarding those that do not match the query conditions — in other words, these indexes are lossy. Because a BRIN index is very small, scanning the index adds little overhead compared to a sequential scan, but may avoid scanning large parts of the table that are known not to contain matching tuples.
https://www.postgresql.org/docs/10/brin-intro.html

Reply:
Django has a postgres-specific BRIN index type that has fascinating potential. But I guess that's not actually getting put in the field definition because it's added automatically as a ForeignKey? It's not a problem for the DB indexes to not align with the ORM?

Reply:
So I really wanted to do this when the new table was *empty* to make the migration faster.  Given our plans to investigate other DB backends in the future, I'd probably shy away from more postgres-specific types if we can avoid it.

Also, in my testing here, (as you can see from our zuul CI which tests for missing migrations), running `makemigrations` seems just fine with this (Django doesn't notice the difference in the index type).